webrender/src/tiling.rs

```diff
@@ -986,34 +1011,45 @@ impl RenderTarget for ColorRenderTarget {
 
-                        // TODO(gw): This should always be fine for now, since the texture
-                        // atlas grows to 4k. However, it won't be a problem soon, once
-                        // we switch the texture atlas to use texture layers!
-                        let textures = BatchTextures {
-                            colors: ctx.prim_store.get_color_textures(prim_metadata),
-                        };
+                        // todo(gw): avoid / recycle this allocation...
+                        let mut instances = Vec::new();
 
-                        debug_assert!(textures.colors[0] != SourceTexture::Invalid);
-                        debug_assert!(self.text_run_textures.colors[0] == SourceTexture::Invalid ||
-                                      self.text_run_textures.colors[0] == textures.colors[0]);
-                        self.text_run_textures = textures;
+                        let font_size_dp = text.logical_font_size.scale_by(ctx.device_pixel_ratio);
 
                         let instance = SimplePrimitiveInstance::new(prim_address,
                                                                     render_tasks.get_task_index(&task.id, pass_index),
                                                                     RenderTaskIndex(0),
                                                                     PackedLayerIndex(0),
                                                                     0);     // z is disabled for rendering cache primitives
 
-                        for glyph_index in 0..text.glyph_instances.len() {
-                            self.text_run_cache_prims.push(instance.build(glyph_index as i32,
-                                                                          text.resource_address.0));
+                        let texture_id = ctx.resource_cache.get_glyphs(text.font_key,
+                                                                       font_size_dp,
+                                                                       text.color,
+                                                                       &text.glyph_instances,
+                                                                       text.render_mode,
+                                                                       text.glyph_options, |index, handle| {
+                            let uv_address = handle.as_int(gpu_cache);
+                            instances.push(instance.build(index as i32, 0, uv_address));
+                        });
+
+                        if texture_id != SourceTexture::Invalid {
+                            let textures = BatchTextures {
+                                colors: [texture_id, SourceTexture::Invalid, SourceTexture::Invalid],
+                            };
+
+                            self.text_run_cache_prims.extend_from_slice(&instances);
+
+                            debug_assert!(textures.colors[0] != SourceTexture::Invalid);
```

Comment:
assertion looks redundant? given that we are in `texture_id != SourceTexture::Invalid` branch